### PR TITLE
Documented new GitLab and Bitbucket webhooks

### PR DESCRIPTION
**Pull request by oatmealraisin**
@ahardin-rh @gabemontero ptal? Does this content warrant a different place to display the info?

### CODE REVIEW

File: dev_guide/builds/triggering_builds.adoc [hunk 1]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+type: "GitLab"
+gitlab:
+  secret: "secret101"
+----
+
+The payload URL is returned as the GitLab Webhook URL by the `describe` command
+(see xref:describe-buildconfig[Displaying Webhook URLs]), and is structured as follows:
+
+----
+http://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/bitbucket
```
Comment (gabemontero) [inline dev_guide/builds/triggering_builds.adoc]: says `bitbucket` at the end ... should say `gitlab`

File: dev_guide/builds/triggering_builds.adoc [hunk 2]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+a trigger definition YAML within the `BuildConfig`:
+
+[source,yaml]
+----
+type: "GitLab"
+gitlab:
+  secret: "secret101"
+----
+
+The payload URL is returned as the GitLab Webhook URL by the `describe` command
```
Comment (gabemontero) [inline dev_guide/builds/triggering_builds.adoc]: Perhaps clarify that `describe` is a command off of our CLI ... i.e. the `oc` executable.

File: dev_guide/builds/triggering_builds.adoc [hunk 3]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+trigger definition YAML within the `BuildConfig`:
+
+[source,yaml]
+----
+type: "Bitbucket"
+bitbucket:
+  secret: "secret101"
+----
+
+The payload URL is returned as the Bitbucket Webhook URL by the `describe` command
```
Comment (gabemontero) [inline dev_guide/builds/triggering_builds.adoc]: Same suggestion to clarify that `describe` is an option on the CLI, `oc` executable, etc.

File: dev_guide/builds/triggering_builds.adoc [hunk 4]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+
+The `-k` argument is only necessary if your API server does not have a properly
+signed certificate.
+
 [[generic-webhooks]]
 === Generic Webhooks
 
 Generic webhooks are invoked from any system capable of making a web request.
-As with a GitHub webhook, you must specify a secret, which will be part of
+As with other webhooks, you must specify a secret, which will be part of
```
Comment (gabemontero) [inline dev_guide/builds/triggering_builds.adoc]: s/As with other webhooks/As with the other webhooks/

**Review by gabemontero — CHANGES_REQUESTED** [review]
On correction and then a couple of minor suggestions.

**Comment by oatmealraisin** [comment]
@gabemontero I've implemented your suggestions

**Review by gabemontero — APPROVED** [review]
IGTM

**Comment by gabemontero** [comment]
@bparees fyi / ptal

File: dev_guide/builds/triggering_builds.adoc [hunk 5]
```diff
@@ -27,7 +27,10 @@ triggers are available:
 
 Webhook triggers allow you to trigger a new build by sending a request to the
 {product-title} API endpoint. You can define these triggers using
-https://developer.github.com/webhooks/[GitHub webhooks] or Generic webhooks.
+https://developer.github.com/webhooks/[GitHub webhooks],
```
Comment (bparees) [inline dev_guide/builds/triggering_builds.adoc]: s/GitHub webhooks/GitHub/

**Review by bparees — CHANGES_REQUESTED** [review]
one nit and lgtm

File: dev_guide/builds/triggering_builds.adoc [hunk 6]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+. Describe the build configuration to get the webhook URL:
++
+----
+$ oc describe bc <name>
+----
+. Copy the webhook URL.
+. Follow the https://docs.gitlab.com/ce/user/project/integrations/webhooks.html#webhooks[GitLab setup instructions]
+to paste the webhook URL into your GitLab repository settings.
+
+Given a file containing a valid JSON payload, you can manually trigger the
```
Comment (php-coder) [inline dev_guide/builds/triggering_builds.adoc]: The command below is expecting the file with particular name. I suggest to specify file name here explicitly.

File: dev_guide/builds/triggering_builds.adoc [hunk 7]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+webhook via `curl`:
+
+----
+$ curl -H "X-GitLab-Event: Push Hook" -H "Content-Type: application/json" -k -X POST --data-binary @gitlub_payload_file.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/gitlab
```
Comment (php-coder) [inline dev_guide/builds/triggering_builds.adoc]: s/gitlub_payload_file/gitlab_payload_file/

File: dev_guide/builds/triggering_builds.adoc [hunk 8]
```diff
@@ -90,11 +93,97 @@ $ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST
...
+----
+. Copy the webhook URL.
+. Follow the https://confluence.atlassian.com/bitbucket/manage-webhooks-735643732.html[Bitbucket setup instructions]
+to paste the webhook URL into your Bitbucket repository settings.
+
+Given a file containing a valid JSON payload, you can manually trigger the
+webhook via `curl`:
+
+----
+$ curl -H "X-Event-Key: repo:push" -H "Content-Type: application/json" -k -X POST --data-binary @gitlub_payload_file.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/bitbucket
```
Comment (php-coder) [inline dev_guide/builds/triggering_builds.adoc]: s/gitlub_payload_file/bitbucket_payload_file/

Or even better -- we can use similar `payload_file.json` name in all examples.

**Comment by oatmealraisin** [comment]
@php-coder @bparees I've implemented your changes!

**Comment by bparees** [comment]
@openshift/team-documentation ptal.  @oatmealraisin please link the PR that implements this feature so the docs team knows when they can merge it.

**Comment by oatmealraisin** [comment]
[Related Origin PR](https://github.com/openshift/origin/pull/13389), thanks @bparees!

File: dev_guide/builds/triggering_builds.adoc [hunk 9]
```diff
@@ -27,7 +27,10 @@ triggers are available:
 
 Webhook triggers allow you to trigger a new build by sending a request to the
 {product-title} API endpoint. You can define these triggers using
-https://developer.github.com/webhooks/[GitHub webhooks] or Generic webhooks.
+https://developer.github.com/webhooks/[GitHub],
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: we need to add link: in front of all links to external websites per our [doc guidelines](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#links-hyperlinks-and-cross-references) for example:

    link:https://developer.github.com/webhooks/[GitHub]

File: dev_guide/builds/triggering_builds.adoc [hunk 10]
```diff
@@ -80,11 +83,97 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+$ curl -H "X-GitHub-Event: push" -H "Content-Type: application/json" -k -X POST --data-binary @payload.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/github
+----
+
+The `-k` argument is only necessary if your API server does not have a properly
+signed certificate.
+
+[[gitlab-webhooks]]
+=== GitLab Webhooks
+
+https://docs.gitlab.com/ce/user/project/integrations/webhooks.html[GitLab webhooks]
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: link:https://docs.gitlab.com/ce/user/project/integrations/webhooks.html[GitLab webhooks]

File: dev_guide/builds/triggering_builds.adoc [hunk 11]
```diff
@@ -80,11 +83,97 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+
+To configure a GitLab Webhook:
+
+. Describe the build configuration to get the webhook URL:
++
+----
+$ oc describe bc <name>
+----
+. Copy the webhook URL.
+. Follow the https://docs.gitlab.com/ce/user/project/integrations/webhooks.html#webhooks[GitLab setup instructions]
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: link:https://docs.gitlab.com/ce/user/project/integrations/webhooks.html#webhooks[GitLab setup instructions]

File: dev_guide/builds/triggering_builds.adoc [hunk 12]
```diff
@@ -80,11 +83,97 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+$ curl -H "X-GitLab-Event: Push Hook" -H "Content-Type: application/json" -k -X POST --data-binary @payload.json https://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/gitlab
+----
+
+The `-k` argument is only necessary if your API server does not have a properly
+signed certificate.
+
+[[bitbucket-webhooks]]
+=== Bitbucket Webhooks
+
+https://confluence.atlassian.com/bitbucket/manage-webhooks-735643732.html[Bitbucket webhooks]
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: link:https://confluence.atlassian.com/bitbucket/manage-webhooks-735643732.html[Bitbucket webhooks]

File: dev_guide/builds/triggering_builds.adoc [hunk 13]
```diff
@@ -80,11 +83,97 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+http://<openshift_api_host:port>/oapi/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/bitbucket
+----
+
+To configure a Bitbucket Webhook:
+
+. Describe the build configuration to get the webhook URL:
++
+----
+$ oc describe bc <name>
+----
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: Add a new line between the code block and the next step

File: dev_guide/builds/triggering_builds.adoc [hunk 14]
```diff
@@ -80,11 +83,97 @@ Therefore, if you are using a Gogs server, you can define a GitHub webhook
...
+. Copy the webhook URL.
+. Follow the https://confluence.atlassian.com/bitbucket/manage-webhooks-735643732.html[Bitbucket setup instructions]
```
Comment (ahardin-rh) [inline dev_guide/builds/triggering_builds.adoc]: link:

**Comment by ahardin-rh** [comment]
@oatmealraisin Overall, looks good! Just a few minor comments from me. I will mark this [do not merge] to prevent premature merging on the docs end. Thanks so much! :beers:

**Comment by ahardin-rh** [comment]
@oatmealraisin Also, to answer your original question, placement of the information LGTM right now. If this topic continues to expand, we may choose to chunk it up later.

**Comment by oatmealraisin** [comment]
@ahardin-rh I've implemented your changes!

**Comment by ahardin-rh** [comment]
@oatmealraisin LGTM!

**Comment by oatmealraisin** [comment]
@ahardin-rh The source PR was merged, could we merge this PR?

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../dev_guide/builds/triggering_builds.adoc#dev-guide-triggering-builds[Builds -> Triggering Builds]
|Added new sections on xref:../dev_guide/builds/triggering_builds.adoc#gitlab-webhooks[GitLab] and  xref:../dev_guide/builds/triggering_builds.adoc#bitbucket-webhooks[Bitbucket] webhooks.
%

**Pull request merged**